larq/optimizers_v2.py

```diff
@@ -1,10 +1,151 @@
+                raise TypeError(
+                    f"Expected `tf.keras.optimizers.Optimizer` at `pred_opt_pairs[{i}][1]` but got `{type(optimizer)}`."
+                )
+
+        # Type check for default optimizers
+        if default is not None and not isinstance(
+            default, tf.keras.optimizers.Optimizer
+        ):
+            raise TypeError(
+                f"Expected `tf.keras.optimizers.Optimizer` for `default` but got `{type(fallback_optimizer)}`."
+            )
+
+        self.pred_opt_pairs = pred_opt_pairs
+        self.default = default
+
+        self.var_opt_mapping = None
+        self.DEFAULT_OPT_INDEX = -1
+
+    def apply_gradients(self, grads_and_vars, name=None):
+        """Apply gradients to variables for each optimizer.
+        
+        On the first call to `apply_gradients()`, compute the mapping from variables to
+        optimizers and cache it in the `self.var_opt_mapping` dict for serialization and
+        faster access.
+        """
+
+        if self.var_opt_mapping is None:
+            grads_and_vars = list(grads_and_vars)
+            self._compute_var_opt_mapping(grads_and_vars)
+
+        train_ops = [
+            self._get_optimizer(var).apply_gradients([(grad, var)])
+            for (grad, var) in grads_and_vars
+        ]
+
+        return tf.group(*train_ops, name="train_with_group")
+
+    def get_config(self):
+        optimizer_configs = [opt.get_config() for (_, opt) in self.pred_opt_pairs]
+        default_config = self.default.get_config()
+
+        config = {
+            "optimizer_configs": [
+                {"class_name": optimizer_config["name"], "config": optimizer_config}
+                for optimizer_config in optimizer_configs
+            ],
+            "default_config": {
+                "class_name": default_config["name"],
+                "config": default_config,
+            },
+            "var_opt_mapping": self.var_opt_mapping,  # serialized instead of `pred`s
+        }
+        return {**super().get_config(), **config}
+
+    @classmethod
+    def from_config(cls, original_config, custom_objects=None):
+        config = deepcopy(original_config)
+
+        case_optimizer = cls(
+            pred_opt_pairs=[  # `(pred, opt)` tuples
+                (
+                    lambda _: False,  # placeholder callable (`pred` is not serialized)
+                    tf.keras.optimizers.deserialize(  # optimizer `opt`
+                        opt_config, custom_objects=custom_objects
+                    ),
+                )
+                for opt_config in config["optimizer_configs"]
+            ],
+            default=tf.keras.optimizers.deserialize(
+                config["default_config"], custom_objects=custom_objects
+            ),
+        )
+
+        # Since we no longer have the `pred`s, we set the mapping explicitly
+        case_optimizer.var_opt_mapping = config["var_opt_mapping"]
```

Comment:
I think this is OK to do, another way would be to pass it as `**kwargs` to the class and catch it in the `__init__`. I don't have a preference, I think both ways are equally valid.

Reply:
I had that initially but I slightly prefer not having it in the `__init__` since that makes it very clear that this isn't something the user should be doing by hand.